### PR TITLE
C Library support for encryption

### PR DESCRIPTION
**Pull request by dwsteele**

### CODE REVIEW

File: lib/pgBackRest/Storage/Filter/CipherBlock.pm [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+            {name => 'tCipherKey', trace => true},
+            {name => 'strMode', optional => true, default => STORAGE_ENCRYPT, trace => true},
+        );
+
+    # Bless with new class
+    my $self = $class->SUPER::new($oParent);
+    bless $self, $class;
+
+    # Check mode is valid
+    $self->{strMode} = $strMode;
```
[the remark this answers is not in the document]
Comment (dwsteele) [reply lib/pgBackRest/Storage/Filter/CipherBlock.pm]: This is on purpose.  The variable that will actually be used later on is the one that is checked for validity so that the assignment is checked as well.  If an error occurs the value of the variable is not important.

File: lib/pgBackRest/Storage/Filter/CipherBlock.pm [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    my $tBufferRead = '';
+    my $iBufferReadSize = 0;
+
+    do
+    {
+        # Read data
+        my $tCipherBuffer;
+        my $iActualSize = $self->SUPER::read(\$tCipherBuffer, $iSize);
+
+        # If something was read, the process it
```
[the remark this answers is not in the document]
Comment (dwsteele) [reply lib/pgBackRest/Storage/Filter/CipherBlock.pm]: Done.

File: test/lib/pgBackRestTest/Module/Storage/StorageFilterCipherBlockTest.pm [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+        #---------------------------------------------------------------------------------------------------------------------------
+        # Create an unencrypted file
+        executeTest("echo -n '${strFileContent}' | tee ${strFile}");
+
+        $self->testException(
+            sub {new pgBackRest::Storage::Filter::CipherBlock(
+                $oDriver->openRead($strFile), $strCipherType, $tCipherKey, {strMode => BOGUS})},
+                ERROR_ASSERT, 'unknown cipher mode: ' . BOGUS);
+
+        $self->testException(
```
[the remark this answers is not in the document]
Comment (dwsteele) [reply test/lib/pgBackRestTest/Module/Storage/StorageFilterCipherBlockTest.pm]: Hmmm - can't figure it out what it was for.  Removed.

File: test/lib/pgBackRestTest/Module/Storage/StorageFilterCipherBlockTest.pm [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+        my $oEncryptBinFileIo = $self->testResult(
+            sub {new pgBackRest::Storage::Filter::CipherBlock(
+                $oDriver->openRead($strFileBinEncrypt), $strCipherType, $tCipherKey,
+                {strMode => STORAGE_DECRYPT})},
+            '[object]', 'new read encrypted bin file');
+
+        $self->testResult(sub {$oEncryptBinFileIo->read(\$tBuffer, 16777216)}, 16777216, '    read 16777216 bytes');
+        $self->testResult(sub {sha1_hex($tBuffer)}, $strFileBinHash, '    check sha1 same as original');
+        $self->testResult(sub {$oEncryptBinFileIo->close()}, true, '    close');
+
```
[the remark this answers is not in the document]
Comment (dwsteele) [reply test/lib/pgBackRestTest/Module/Storage/StorageFilterCipherBlockTest.pm]: Added.

File: test/lib/pgBackRestTest/Module/Storage/StorageFilterCipherBlockTest.pm [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+            '[object]', 'write file to be read by openssl');
+
+        $self->testResult(sub {$oEncryptFileIo->write(\$tBuffer)}, 8, '    write 8 bytes');
+        $self->testResult(sub {$oEncryptFileIo->close()}, true, '    close');
+
+        executeTest(
+            "openssl enc -d -k ${tCipherKey} -md sha1 -aes-256-cbc -in ${strFileEncrypt} -out ${strFile}");
+
+        $self->testResult(sub {${$self->storageTest()->get($strFile)}}, $strFileContent, '    check content same as original');
+    }
```
[the remark this answers is not in the document]
Comment (dwsteele) [reply test/lib/pgBackRestTest/Module/Storage/StorageFilterCipherBlockTest.pm]: Actually, no encrypted file should ever be 0 bytes.  So this was a problem in code.  0 byte files are now properly encrypted and decrypted and I have added tests.  Also added a test to error when decrypting an empty file.

Everything worked as it was (and your test worked), but this would have broken compatibility with openssl for 0 byte files.  See 0fe8aac05b58e6ccd37.

**Comment by dwsteele** [comment]
Also, I removed the MEM_CONTEXT blocks from places where mem contexts were not being used.  Originally they were put in "just in case" some mem context operations happened but for high performance code like this I don't think it makes sense.

File: lib/pgBackRest/Storage/Filter/CipherBlock.pm (hunk 1 above)
[the remark this answers is not in the document]
Comment (cmwshang) [reply lib/pgBackRest/Storage/Filter/CipherBlock.pm]: It just seems to me that it would be more efficient to check $strMode and then set $self->{strMode} = $strMode if $strMode it valid.

[the remark this answers is not in the document]
Comment (dwsteele) [reply lib/pgBackRest/Storage/Filter/CipherBlock.pm]: I'm going for correctness here more than efficiency.  In all but the most extreme situations that assignment will run anyway.  The error is an ASSERT - so really it should never be hit at all.
